File: slides/omp/code/sum_thread_2.f90
program sum_parallel
  implicit none
  integer*8, parameter :: &
    n = 100000000
  integer :: i
  real*8 :: s, s_ref
  real*8, allocatable, &
    dimension(:) :: v
  allocate(v(n))
  do i = 1, n
    v(i) = dble(i-1)
  end do
  !$omp parallel do
  do i = 1, n
    !$omp atomic
    s = s + v(i)
  end do
  !$omp end parallel do
  s_ref = 0.5_8*(n*n-n)
  print *, "sum is ", s, &
    " expected ", s_ref
  deallocate(v)
end program sum_parallel

File: slides/omp/code/sum_thread_2.cc
#include <cstdio>
using std::printf;
int main()
{
  const long n = 100000000;
  double *v = new double[n];
  for (int i = 0; i < n; ++i)
    v[i] = (double)i;
  double s = 0.0;
  #pragma omp parallel for
  for (int i = 0; i < n; ++i)
  {
    #pragma omp atomic
    s += v[i];
  }
  double s_ref = 0.5*(n*n-n);
  printf("sum is %f, "
      "expected %f\n", s, s_ref);
  delete [] v;
}
